### DISCUSSION TITLE
How can xstate be used to replace global stores like redux slices ?

### DISCUSSION THREAD

**Discussion by dsathyakumar** [statelyai/xstate#2305] [Q&A]
Hello 👋 

I'd like to know of samples where Xstate can be used to replace redux & its slices. 
As Xstate uses the Actor Model, it ideally can dispatch events to other machines / actors.
So Assuming there is a component A and about 10 instances of component B on the page, is there an example of how these 2 machines can interact (Dispatch & receive events), how do all 10 instances receive & respond to events etc.,

**Comment by ChrisShank** [comment]
Just confirming before I answer that you are thinking that each component will instantiate its own machine instance (by using `useMachine`)?

**Reply by dsathyakumar** [reply]
Yes (as the machine is tied to the component via the `useMachine` hook. or is my understanding wrong ? 🤔

**Reply by ChrisShank** [reply]
`xstate` is interesting because it can be used at the component-level and the app-level (i.e *like* a global store)!

At the component-level (when you use `useMachine`) the machine usually defines the behavior of the component. You are tying the instance of the machine to the lifespan of the component. All prop changes and events emitted from children are sent as events to the machine for processing. The component renders something based on the current `state` of the machine.

At the app-level `xstate` can completely replace any state management solution! It can be used as an single-root store like Redux or a multi-root store like Jotai. The key is defining the instance of the machine so it is alive for the entire lifespan of the app. This is typically done using `useMachine` or `useInterpret` in your root component then passing that service around your app using props or React’s Context. For any component where you want to subscribe to the machine use the `useService` hook.

**Reply by ChrisShank** [reply]
Now going a little deeper, at either level you are not limited to only instantiating one machine per instantiation! A machine can invoke or spawn other machines, which means that you can create a tree of actors that encapsulate a certain behavior. This is where the actor model really comes into play and in my experience has been the most robust way to build out complex behavior/logic, especially at the app-level! In components you can subscribe to individual actors by doing something like this `useActor(state.children.someActor)`. 

So at the app-level this is where `xstate` tends to diverge from (my basic understanding of) Redux. Your app now has two layers: one layer is actor tree that models the behavior of your app, the other layer is the component tree that subscribes to certain parts of the behavior and send events back to it!
